retrieval/transform.go

```diff
@@ -272,6 +274,110 @@ func (d *distribution) Swap(i, j int) {
+// NOTE: buildSummary() EXTENSIVELY mimics the logic of buildHistogram() - consider refactoring (properly).
+func (b *sampleBuilder) buildSummary(
+	ctx context.Context,
+	baseName string,
+	matchLset labels.Labels,
+	samples []record.RefSample,
+) (*metric_pb.DoubleSummaryDataPoint, int64, []record.RefSample, error) {
+	var (
+		consumed       int
+		count, sum     float64
+		resetTimestamp int64
+		lastTimestamp  int64
+		values	       = make([]*metric_pb.DoubleSummaryDataPoint_ValueAtQuantile, 0)
+		skip           = false
+	)
+	// We assume that all series belonging to the summary are sequential. Consume series
+	// until we hit a new metric.
+Loop:
+	for i, s := range samples {
+		e, err := b.series.get(ctx, s.Ref)
+		if err != nil {
+			// Note: This case may or may not trigger the
+			// len(samples) == len(newSamples) test in
+			// manager.go. The important part here is that
+			// we may skip or may not skip any points that
+			// belong to the histogram, but if we don't
+			// the manager safetly advances.
+			return nil, 0, samples, err
+		}
+		if e == nil {
+			// These points were filtered. This seems like
+			// an impossible situation.
+			break
+		}
+		name := e.lset.Get("__name__")
+		// The series matches if it has the same base name, the remainder is a valid summary suffix,
+		// and the labels aside from the `quantile` and __name__ label match up.
+		if !strings.HasPrefix(name, baseName) || !labelsEqual(e.lset, matchLset, "quantile") {
+			break
+		}
+		// This is an edge case of having an adjacent point with the same base name and labels,
+		// but different type.
+		if e.metadata.MetricType != textparse.MetricTypeSummary {
+			break
+		}
+		// In general, a scrape cannot contain the same (set of) series repeatedlty but for different timestamps.
+		// It could still happen with bad clients though and we are doing it in tests for simplicity.
+		// If we detect the same series as before but for a different timestamp, return the summary up to this
+		// series and leave the duplicate time series untouched on the input.
+		if i > 0 && s.T != lastTimestamp {
+			// TODO: counter
+			break
+		}
+		lastTimestamp = s.T
+
+		rt, v := b.series.getResetAdjusted(e, s.T, s.V)
+
+		switch e.suffix {
+		case metricSuffixSum:
+			sum = v
+		case metricSuffixCount:
+			count = v
+			// We take the count series as the authoritative source for the overall reset timestamp.
+			resetTimestamp = rt
+		case "": // Actual values at quantiles.
+			quantile, err := strconv.ParseFloat(e.lset.Get("quantile"), 64)
+
+			if err != nil {
+				consumed++
+				// TODO: increment metric.
+				continue
+			}
+			values = append(values, &metric_pb.DoubleSummaryDataPoint_ValueAtQuantile{
+				Quantile: quantile,
+				Value: v,
+			})
+		default:
+			break Loop
+		}
+		consumed++
+	}
+	// Don't emit a sample if we explicitly skip it or no reset timestamp was set because the
+	// count series was missing.
+	if skip || resetTimestamp == 0 {
```

Comment:
nit: skip is a dead variable. will you remove it here (and in `buildHistogram()` where it is equally dead)?

Reply:
Oh yesss, had totally forgotten about that one.

Reply:
Oh yesss, had totally forgotten about that one.